### ISSUE TITLE
How can we make this work long term?

### ISSUE THREAD

**Issue by physiii** [bigscience-workshop/petals#483]
I do not understand how this can be stable since you are relying on donating memory and compute?

I would love to use this concept in production over AWS but I do not see how this will work long term unless you have some way of bartering.

For example, if I have a 10x more "powerful" computer (cpu/gpu/tpu etc) then you have to give 10 computes for every 1 of my computes.

This way the system is balanced otherwise you are relying on philanthropy which is not stable.

Is there any way to incorporate a memory/compute based economy?

**Comment by tsadpbb** [comment]
Crypto would be my first thought. You get rewarded tokens based on compute and you use tokens to access the network.

**Comment by shkarlsson** [comment]
I agree. Traditional torrenting generally can rely on people donating resources because some hard drive space and network has a very low cost. Donating GPU, which is essentially what happens at petals, doesn't seem to work (judging on the short list of donors over at https://health.petals.dev/

Is there some sort of discussion or project addressing this? Maybe an equivalent to closed torrenting sites that require some minimum donation ratio? Or built in micropayments with crypto as I think @earonesty suggests [here](https://github.com/bigscience-workshop/petals/issues/494#issue-1875861792)?

**Comment by ELigoP** [comment]
I would look from psychological perspective, too: if there would be clear objectives what users are going to do on GPUs (e.g. research, benchmarks etc.), it would attract attention and increase participation of people with GPUs.

E.g. with SETI@home, folding@home there is clearly a clear cause to donate GPU time.

If the cause would be to train LLM for specific tasks, train private+RAG LLMs, benchmark existing LLMs that will benefit individual owners of clusters - that would help.

**Comment by bitnom** [comment]
I'm trying to make some strides in this vein. I need better metrics though. I need to know some inference speed stats for various models and hardware on a network. We should be able to revive this effort depending on the average performance of modern models on a healthy network. Currently, nobody is contributing compute though. The one healthy node running bloom at 1 token/sec isn't gonna cut it.

**Comment by bitnom** [comment]
nvm my last question. Upon doing a deep-dive into bottlenecks idk if it's possible to make a viable scaled inference system like this. Maybe MoE makes performance acceptable. I'm not sure but I think it would be easier to focus on economic model alternatives.

I would be curious to know what the maintainer's vision of the perfect petals network would look like, and what kind of performance they might expect from it. I'm speaking solely on inference, not training or fine-tuning.

**Comment by bitnom** [comment]
See https://github.com/learning-at-home/hivemind/issues/647 for a shot in the dark

**Comment by echeloni11** [comment]
> nvm my last question. Upon doing a deep-dive into bottlenecks idk if it's possible to make a viable scaled inference system like this. Maybe MoE makes performance acceptable. I'm not sure but I think it would be easier to focus on economic model alternatives.
> 
> I would be curious to know what the maintainer's vision of the perfect petals network would look like, and what kind of performance they might expect from it. I'm speaking solely on inference, not training or fine-tuning.

@bitnom  Hi, do you mind sharing some thoughts about the bottlenecks that you mentioned here? I'm also looking at this project and find it interesting. When you talk about "viable scaled inference system", do you have in mind something more like a serving system (i.e. the system serves a certain given model and deal with many concurrent and small requests)? I'm somehow thinking about a more "throughput oriented" scenario rather than latency, where the system gathers resources and each provider can submit their inference jobs where they have batched inputs (e.g. a lab can do inference experiments on larger models in this system), and in this scenario maybe the cost of communication may be hidden by the computation in each pipeline stage since the inputs are batched, and the throughput may be satisfying.

But that's just some initial thoughts and I'm really new to this area and possibly there are things I failed to consider. Would appreciate any thoughts, Thanks!

**Comment by aodalt03-a11y** [comment]
This is the exact question we started Iriath Grid around — betting that a large enough pool of ordinary volunteer machines can make real training runs happen without a data center, long-term and sustainably. Early days (real bugs, real hardware limits), but if the long-term angle interests you as much as the tech, would love to talk or have you run a worker: https://github.com/aodalt03-a11y/iriath-grid-worker
